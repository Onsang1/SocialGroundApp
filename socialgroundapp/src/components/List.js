import React, { useState } from 'react';
import PropTypes from 'prop-types';
import userProfile from '../icons/userProfile.png';
import { updateFollow } from '../api/mock_api';

function Follow({ follower, following }) {
  const [FollowStatus, setFollowStatus] = useState(false);
  const handleOnClick = async (e) => {
    e.preventDefault();
    const action = FollowStatus === false ? 'follow' : 'unfollow';
    await updateFollow(follower, following, action);
    setFollowStatus(!FollowStatus);
  };

  return (
    <button className="submit-button-find" type="submit" onClick={handleOnClick}>
      <div className="button-label">
        {FollowStatus ? 'Unfollow' : 'Follow'}
      </div>
    </button>
  );
}

function List({
  listType, current, users, input,
}) {
  let data = '';
  if (listType === 'suggestion') {
    const suggested = users.filter((el) => {
      if (current.suggested?.includes(el.id)) {
        return el;
      }
      return null;
    });
    const filteredData = users.filter((el) => {
    // if no input the return the original
      if (input === '') {
        return el;
      } // return the item which contains the user input
      return el.username.toLowerCase().includes(input);
    });
    data = (input === '') ? suggested : filteredData;
  } else if (listType === 'followers') {
    const followers = users.filter((el) => {
      if (current.followers?.includes(el.id)) {
        return el;
      }
      return null;
    });
    data = followers;
  } else if (listType === 'following') {
    const following = users.filter((el) => {
      if (current.following?.includes(el.id)) {
        return el;
      }
      return null;
    });
    data = following;
  }
  return (
    <ul>
      {data.map((user) => (
        <li key={user.id}>
          <div className="user-info-area">
            <div className="left">
              <img src={userProfile} className="profile-icon" alt="" />
            </div>
            <div className="middle">
              <p className="username">
                {user.username}
              </p>
            </div>
            <div className="right">
              <Follow
                following={user}
                follower={current}
              />
            </div>
          </div>
        </li>
      ))}
    </ul>

  );
}

Follow.propTypes = {
  follower: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    username: PropTypes.string.isRequired,
    email: PropTypes.string,
    following: PropTypes.arrayOf(PropTypes.number),
    followers: PropTypes.arrayOf(PropTypes.number),
    suggested: PropTypes.arrayOf(PropTypes.number),
  }).isRequired,
  following: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    username: PropTypes.string.isRequired,
    email: PropTypes.string,
    following: PropTypes.arrayOf(PropTypes.number),
    followers: PropTypes.arrayOf(PropTypes.number),
    suggested: PropTypes.arrayOf(PropTypes.number),
  }).isRequired,
};

List.propTypes = {
  input: PropTypes.string.isRequired,
  current: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    username: PropTypes.string.isRequired,
    email: PropTypes.string,
    following: PropTypes.arrayOf(PropTypes.number),
    followers: PropTypes.arrayOf(PropTypes.number),
    suggested: PropTypes.arrayOf(PropTypes.number),
  }).isRequired,
  users: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    username: PropTypes.string.isRequired,
    email: PropTypes.string,
    password: PropTypes.string,
    following: PropTypes.arrayOf(PropTypes.number),
    followers: PropTypes.arrayOf(PropTypes.number),
    postCount: PropTypes.number,
  }).isRequired).isRequired,
  listType: PropTypes.string.isRequired,
  follower: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    username: PropTypes.string.isRequired,
    email: PropTypes.string,
    following: PropTypes.arrayOf(PropTypes.number),
    followers: PropTypes.arrayOf(PropTypes.number),
    suggested: PropTypes.arrayOf(PropTypes.number),
  }).isRequired,
  following: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    username: PropTypes.string.isRequired,
    email: PropTypes.string,
    following: PropTypes.arrayOf(PropTypes.number),
    followers: PropTypes.arrayOf(PropTypes.number),
    suggested: PropTypes.arrayOf(PropTypes.number),
  }).isRequired,
};

export default List;
