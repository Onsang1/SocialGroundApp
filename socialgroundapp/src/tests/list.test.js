/**
* @jest-environment jsdom
*/

import React from 'react';
// import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import '@testing-library/jest-dom';
import { test, expect } from '@jest/globals';
import List from '../components/List';

const mockedUsers = [
  {
    id: 1,
    name: 'Bret Watermelon',
    username: 'Bret',
    email: 'bwatermelon@gmail.com',
    password: 'bret123',
    following: [
      2,
      4,
    ],
    followers: [
      3,
    ],
    postCount: 1,
  },
  {
    id: 2,
    name: 'Ervin Howell',
    username: 'Antonette',
    email: 'Shanna@melissa.tv',
    password: 'ant123',
    following: [],
    followers: [],
    postCount: 0,
  }];

const mockCurrUser = [
  {
    id: 1,
    name: 'Lena Sheng',
    username: 'xinyuesh',
    email: 'xinyuesh@gmail.com',
    following: [2, 4],
    followers: [3],
    suggested: [1, 2, 3],
  },
];

test('test to see if list components are rendered', async () => {
  const tree = renderer.create(<List
    input={[]}
    current={mockCurrUser}
    users={Array.from(mockedUsers)}
    listType="suggestion"
  />);
  expect(tree).toMatchSnapshot();
//   const linkElement = screen.getByText('Friends');
//   expect(linkElement).toBeInTheDocument();
//   expect(getByText(/Find Friends on Social/i)).toBeInTheDocument();
  // expect(Follow(follower={}, following={})).toHaveBeenCalled();
});
