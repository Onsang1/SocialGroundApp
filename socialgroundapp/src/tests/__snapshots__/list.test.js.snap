// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test to see if list components are rendered 1`] = `<ul />`;
